Subscribers

IdanTo1 opened the pull request.

IdanTo1 committed:
change all messages methods to public

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

IdanTo1 committed:
change all messages methods to public - one more

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

IdanTo1 committed:
change all messages methods to public - one more

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

IdanTo1 committed:
implement all subscribers

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

Hroy02 commented on src/main/java/bgu/spl/mics/application/subscribers/M.java:
```diff
@@ -1,24 +1,103 @@
+            // check whether Q acquired the gadget, and allocate all relevant details
+            gadgetAvailableObject = gadgetAvailableFuture.get(); //M and Q uses the same gadgetObject
+            if (!gadgetAvailableObject.isGadgetExists()) return;
+            String gadget = gadgetAvailableObject.getGadget();
```
If the gadget doesn't exist you should probably release the agents

Hroy02 commented on src/main/java/bgu/spl/mics/application/subscribers/M.java:
```diff
@@ -1,24 +1,103 @@
+            }
+            // signal Moneypenny she should send the agents.
+            agentsAvailableObject.sendMission();
 
```
Shouldn't there be some sort of signaling to Moneypenny that something changed? maybe a `agentsAvailableObject.notifyAll()` or something

Hroy02 commented on src/main/java/bgu/spl/mics/application/subscribers/Moneypenny.java:
```diff
@@ -10,15 +16,34 @@
-		super("Change_This_Name");
-		// TODO Implement this
+	public Moneypenny(int serial) {
+		super(("Moneypenny" + serial));
```
redundant parentheses

IdanTo1 replied:
I thought it is clearer this way

Hroy02 commented on src/main/java/bgu/spl/mics/application/subscribers/Moneypenny.java:
```diff
@@ -10,15 +16,34 @@
+			List<String> agentsSerials = result.getAgentsSerials();
+			java.util.Collections.sort(agentsSerials);
+			result.setMoneypennySerial(_serial);
+			if(squad.getAgents(agentsSerials)) {
+				result.setAgentsNames(squad.getAgentsNames(agentsSerials));
+			}
```
getAgents acquires the agents. You will not be able to get them twice

IdanTo1 replied:
I'm calling `getAgents` only once.. pay attention the second one is `getAgentsNames`

Hroy02 commented on src/main/java/bgu/spl/mics/application/subscribers/Moneypenny.java:
```diff
@@ -10,15 +16,34 @@
+			complete(e, result);
+			while (!result.isSendMission() || !result.isTerminateMission()) {
+				try {e.wait();} catch (InterruptedException ignored) {}
+			}
```
You should probably wait on the result, since M creates the event without saving it

Hroy02 reviewed (changes requested):
Good work! some bugs but they are easy fixes

IdanTo1 committed:
fix due to CR
1. add terminateMission in case there is no gadget.
2. notify Moneypenny at all relevant occasions.
3. Moneypenny is waiting on result and not the event as it is the common object

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

Hroy02 reviewed (approved):
Great! You can merge
